scripts/tests/test_mnist_experiment.py

```
@@ -0,0 +1,170 @@
+            self.assertFalse(output_directory.exists())
+            self.assertEqual(list(root.iterdir()), [])
+
+    def test_existing_run_is_never_overwritten(self) -> None:
+        """A caller must choose a new output path for every generation."""
+        with TemporaryDirectory() as temporary_directory:
+            output_directory = Path(temporary_directory) / "run"
+            output_directory.mkdir()
+            sentinel = output_directory / "run.json"
+            sentinel.write_text("previous\n", encoding="utf-8")
+
+            with self.assertRaisesRegex(ValueError, "already exists"), _staged_run_directory(output_directory):
```

Ruff (0.16.1)

[warning] Use `pytest.raises` instead of unittest-style `assertRaisesRegex`

Replace `assertRaisesRegex` with `pytest.raises`

(PT027)